keps/sig-storage/1472/README.md

```diff
@@ -0,0 +1,1416 @@
+can be used to build and run a custom scheduler where the desired
+policy is compiled into the scheduler binary.
+
+## Proposal
+
+### Caching remaining capacity via the API server
+
+The Kubernetes scheduler cannot talk directly to the CSI drivers to
+retrieve capacity information because CSI drivers typically only
+expose a local Unix domain socket and are not necessarily running on
+the same host as the scheduler(s).
+
+The key approach in this proposal for solving this is to gather
+capacity information, store it in the API server, and then use that
+information in the scheduler. That information then flows
+through different components:
+1. storage backend
+2. CSI driver
+3. Kubernetes-CSI sidecar
+4. API server
+5. Kubernetes scheduler
+
+The first two are driver specific. The sidecar will be provided by
+Kubernetes-CSI, but how it is used is determined when deploying the
+CSI driver. Steps 3 to 5 are explained below.
+
+### Gathering capacity information
+
+A sidecar, external-provisioner in this proposal, will be extended to
+handle the management of the new objects. This follows the normal
+approach that integration into Kubernetes is managed as part of the
+CSI driver deployment, ideally without having to modify the CSI driver
+itself. This will work without changes for some storage systems while
+others may have to support [a new CSI API
+call](#with-central-controller-generic-solution).
+
+### Pod scheduling
+
+The Kubernetes scheduler watches the capacity information and excludes
+nodes with insufficient remaining capacity when it comes to making
+scheduling decisions for a pod which uses persistent volumes with
+delayed binding. If the driver does not indicate that it supports
+capacity reporting, then the scheduler proceeds just as it does now,
+so nothing changes for existing CSI driver deployments.
+
+## Design Details
+
+### API
+
+#### CSIStorageCapacity
+
+```
+// CSIStorageCapacity stores the result of one CSI GetCapacity call for one
+// driver, one topology segment, and the parameters of one storage class.
+type CSIStorageCapacity struct {
+	metav1.TypeMeta
+	// Standard object's metadata. The name has no particular meaning and just has to
+	// meet the usual requirements (length, characters, unique). To ensure that
+	// there are no conflicts with other CSI drivers on the cluster, the recommendation
+	// is to use csisc-<uuid>.
+	//
+	// Objects are not namespaced.
+	//
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta
+
+    // Spec contains the properties that describe one capacity value.
+    Spec CSIStorageCapacitySpec
+}
+
+// CSIStorageCapacitySpec contains the properties that describe one capacity value.
+type CSIStorageCapacitySpec struct {
+	// The CSI driver that provides the storage.
+	// This must be the string returned by the CSI GetPluginName() call.
+	DriverName string
```

Comment:
Strictly speaking, the driver name is redundant here. It was a left-over from the time when storage class was optional (CSI ephemeral volumes, the optimization for "storage class does not affect capacity").

But after having thought a bit about it, I think it is worthwhile keeping because it enables the listing of all objects for one driver, something that the external-provisioner has to do.